implementations/rust/ockam/ockam_api/src/kafka/kafka_portal_worker.rs

```diff
@@ -0,0 +1,423 @@
+use crate::kafka::decoder::KafkaDecoder;
+use crate::kafka::encoder::KafkaEncoder;
+use crate::kafka::kafka_inlet_map::KafkaInletMap;
+use crate::kafka::protocol_aware::ProtocolState;
+use ockam::Context;
+use ockam_core::{
+    errcode::{Kind, Origin},
+    Address, AllowAll, Decodable, Encodable, Error, LocalInfo, LocalMessage, Route, Routed,
+    TransportMessage, Worker,
+};
+use ockam_transport_tcp::{PortalMessage, MAX_PAYLOAD_SIZE};
+use std::sync::Arc;
+
+///by default kafka supports up to 1MB messages
+pub(crate) const MAX_KAFKA_MESSAGE_SIZE: usize = 16 * 1024 * 1024;
+
+enum Receiving {
+    Requests,
+    Responses,
+}
+
+///Acts like a relay for messages between tcp inlet and outlet for both directions.
+/// It's meant to be created by the portal listener.
+///
+/// This instance manage both streams inlet and outlet in two different workers, one dedicated
+/// to the requests (inlet=>outlet) the other for the responses (outlet=>inlet).
+/// since every kafka message is length-delimited every message is read and written
+/// through a framed encoder/decoder.
+///
+/// ```text
+/// ┌────────┐  decoder    ┌─────────┐  encoder    ┌────────┐
+/// │        ├────────────►│ Kafka   ├────────────►│        │
+/// │        │             │ Request │             │        │
+/// │  TCP   │             └─────────┘             │  TCP   │
+/// │ Inlet  │             ┌─────────┐             │ Outlet │
+/// │        │  encoder    │ Kafka   │   decoder   │        │
+/// │        │◄────────────┤ Response│◄────────────┤        │
+/// └────────┘             └─────────┘             └────────┘
+///```
+pub(crate) struct KafkaPortalWorker {
+    other_worker_address: Address,
```

Comment:
Would be nice to have some comment about what is the "other" worker here. Also, what is the relation between them, is one of them responsible for shutting down the other?

Reply:
yep, adding comments!